### PR TITLE
[Don't merge][Review] A solution to prevent zombie containers locally and in CI

### PR DESCRIPTION
**Pull request by larroy**
## Description ##
This PR adds mechanisms on the build scripts to cleanup docker containers when there is a cancellation, either from the user by sending SIGINT / SIGTERM or when the process inside the container ends with an error.

It also propagates the environment variables from Jenkins which are used by the process tree killer to identify runaway processes so processes inside the container are killed when the Jenkins job is stopped.

With this patch we catch SIGINT and SIGTERM and install a handler on atexit which cleans (stops) docker containers which were created during execution of the build.

We also switch to python docker API for better management of running containers.

@aaronmarkham @marcoabreu @lebeg @KellenSunderland

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by larroy** [comment]
@mxnet-label-bot: [pr-awaiting-review]

File: ci/build.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
 """
 
 __author__ = 'Marco de Abreu, Kellen Sunderland, Anton Chernov, Pedro Larroy'
-__version__ = '0.1'
+__version__ = '0.8'
```
Comment (zhreshold) [inline ci/build.py]: why bumping to 0.8 directly?

Comment (larroy) [reply ci/build.py]: because it's getting close to 1.0

Comment (zhreshold) [reply ci/build.py]: ok

File: ci/build.py [hunk 2]
```diff
@@ -91,40 +192,24 @@ def build_docker(platform: str, docker_binary: str, registry: str, num_retries:
...
+    #
+    cmd = [docker_binary, "build",
+           "-f", get_dockerfile(platform),
+           "--build-arg", "USER_ID={}".format(os.getuid()),
+           "--build-arg", "GROUP_ID={}".format(os.getgid())]
+    if use_cache:
+        cmd.extend(["--cache-from", tag,])
+    cmd.extend(["-t", tag, get_dockerfiles_path()])
+
+    @retry(subprocess.CalledProcessError, tries=num_retries)
```
Comment (KellenSunderland) [inline ci/build.py]: Cool but should be in a different PR.  Reverting your zombie containers fix would mean we'd have to re-implement your retry refactor and vice-versa.

File: ci/build.py [hunk 3]
```diff
@@ -158,63 +246,142 @@ def default_ccache_dir() -> str:
...
-        ret = call(into_cmd)
-
-    if not dry_run and not interactive and ret != 0:
-        logging.error("Running of command in container failed (%s):\n%s\n", ret, cmd)
-        logging.error("You can get into the container by adding the -i option")
-        raise subprocess.CalledProcessError(ret, cmd)
-
-    return docker_run_cmd
+    docker_cmd_list.extend(command)
+    docker_cmd = ' \\\n\t'.join(docker_cmd_list)
```
Comment (KellenSunderland) [inline ci/build.py]: Would you be open to not putting these on different lines with \ ?  I've had to copy and paste these and each time I have to paste it into a text editor, removed the \ and newlines.

File: ci/build.py [hunk 4]
```diff
@@ -158,63 +246,142 @@ def default_ccache_dir() -> str:
...
                '--shm-size={}'.format(shared_memory_size),
                '-v', "{}:/work/mxnet".format(mx_root),  # mount mxnet root
                '-v', "{}:/work/build".format(local_build_folder),  # mount mxnet/build for storing build artifacts
                '-v', "{}:/work/ccache".format(local_ccache_dir),
                '-u', '{}:{}'.format(os.getuid(), os.getgid()),
-               '-e', 'CCACHE_MAXSIZE={}'.format(CCACHE_MAXSIZE),
                '-e', 'CCACHE_TEMPDIR=/tmp/ccache',  # temp dir should be local and not shared
                '-e', "CCACHE_DIR=/work/ccache",  # this path is inside the container as /work/ccache is mounted
                '-e', "CCACHE_LOGFILE=/tmp/ccache.log",  # a container-scoped log, useful for ccache verification.
+               '-ti',
```
Comment (KellenSunderland) [inline ci/build.py]: Great change!  Not specific to zombie containers, should be in its own PR.

File: ci/build.py [hunk 5]
```diff
@@ -383,7 +604,8 @@ def use_cache():
 
 ./build.py -a
 
-    Builds for all platforms and leaves artifacts in build_<platform>
+    Builds for all platforms and leaves artifacts in build_<platform>. **WARNING** it performs git
```
Comment (KellenSunderland) [inline ci/build.py]: See above comment on this.

File: ci/build.py [hunk 6]
```diff
@@ -284,7 +480,7 @@ def script_name() -> str:
                         default=1,
                         type=int)
 
-    parser.add_argument("-c", "--cache", action="store_true",
+    parser.add_argument("--no-cache", action="store_true",
```
Comment (KellenSunderland) [inline ci/build.py]: Can you give a bit more detail about what this change does?  What's its relation to zombie containers?

File: ci/build.py (hunk 1 above)
Comment (KellenSunderland) [reply ci/build.py]: I'd suggest 0.2.

File: ci/build.py [hunk 7]
```diff
@@ -37,17 +37,118 @@
...
+    """Clean repo and subrepos, update subrepos"""
+    logging.info("cleaning up repository")
+    with remember_cwd():
+        os.chdir(get_mxnet_root())
+        check_call(['git', 'clean', '-ffdx'])
+        check_call(['git', 'submodule', 'foreach', '--recursive', 'git', 'clean', '-ffdx'])
+        check_call(['git', 'submodule', 'update', '--recursive', '--init'])
+
+
+def get_dockerfiles_path():
```
Comment (KellenSunderland) [inline ci/build.py]: Maybe a constant instead of a function?

File: ci/build.py [hunk 8]
```diff
@@ -37,17 +37,118 @@
...
+    return "docker"
+
+
+def get_platforms(path: Optional[str] = get_dockerfiles_path()):
```
Comment (KellenSunderland) [inline ci/build.py]: Return val.

File: ci/build.py [hunk 9]
```diff
@@ -158,63 +246,142 @@ def default_ccache_dir() -> str:
...
+        #############################
+        #
+        signal.pthread_sigmask(signal.SIG_BLOCK, {signal.SIGINT, signal.SIGTERM})
+        if nvidia_runtime:
+            runtime='nvidia'
+        else:
+            # runc is default (docker info | grep -i runtime)
+            runtime=None
+
+        container = docker_client.containers.run(
```
Comment (KellenSunderland) [inline ci/build.py]: Really like this

File: ci/build.py [hunk 10]
```diff
@@ -158,63 +246,142 @@ def default_ccache_dir() -> str:
...
+            tag,
+            runtime=runtime,
+            detach=True,
+            command=command,
+            #auto_remove=True,
```
Comment (KellenSunderland) [inline ci/build.py]: Commented code needed?

File: ci/build.py [hunk 11]
```diff
@@ -158,63 +246,142 @@ def default_ccache_dir() -> str:
...
+        # If the previous call is interrupted then it's possible that the container is not cleaned up
+        # We avoid by masking the signals temporarily
+        cleanup.add_container(container)
+        signal.pthread_sigmask(signal.SIG_UNBLOCK, {signal.SIGINT, signal.SIGTERM})
+        #
+        #############################
+
+        stream = container.logs(stream=True, stdout=True, stderr=True)
+        sys.stdout.flush()
+        for chunk in stream:
```
Comment (KellenSunderland) [inline ci/build.py]: Would it be possible to put this in a helper, and describe why it's needed?

File: ci/build.py [hunk 12]
```diff
@@ -226,19 +393,48 @@ def load_docker_cache(tag, docker_registry) -> None:
...
 
-    def script_name() -> str:
-        return os.path.split(sys.argv[0])[1]
 
+def main() -> int:
+    logging.getLogger().setLevel(logging.INFO)
+    logging.getLogger("requests").setLevel(logging.WARNING)
     logging.basicConfig(format='{}: %(asctime)-15s %(message)s'.format(script_name()))
 
+    log_environment()
```
Comment (KellenSunderland) [inline ci/build.py]: I'd make this change in a different PR.

File: ci/build.py [hunk 13]
```diff
@@ -312,31 +521,37 @@ def use_cache():
...
         if command:
-            container_run(platform=platform, docker_binary=docker_binary, shared_memory_size=shared_memory_size,
+            ret = container_run(platform=platform, nvidia_runtime=args.nvidiadocker, shared_memory_size=shared_memory_size,
                           command=command, docker_registry=args.docker_registry,
-                          local_ccache_dir=args.ccache_dir, interactive=args.interactive)
+                          local_ccache_dir=args.ccache_dir, interactive=args.interactive, cleanup=cleanup)
         elif args.print_docker_run:
-            print(container_run(platform=platform, docker_binary=docker_binary, shared_memory_size=shared_memory_size,
-                                command=[], dry_run=True, docker_registry=args.docker_registry, local_ccache_dir=args.ccache_dir))
+            ret = container_run(platform=platform, nvidia_runtime=args.nvidiadocker, shared_memory_size=shared_memory_size,
```
Comment (KellenSunderland) [inline ci/build.py]: Is the cleanup param optional?  It's not filled here.

File: ci/build.py [hunk 14]
```diff
@@ -346,15 +561,21 @@ def use_cache():
...
-            plat_buildir = os.path.join(get_mxnet_root(), build_platform)
+            build_platform = "build_{}".format(platform)
+            plat_buildir = os.path.abspath(os.path.join(get_mxnet_root(), '..',
+                                                        "mxnet_{}".format(build_platform)))
+            if os.path.exists(plat_buildir):
+                logging.warning("{} already exists, skipping".format(plat_buildir))
+                continue
+            command = ["/work/mxnet/ci/docker/runtime_functions.sh", build_platform]
+            # BUILD in docker
+            container_run(platform=platform, nvidia_runtime=args.nvidiadocker, shared_memory_size=shared_memory_size,
```
Comment (KellenSunderland) [inline ci/build.py]: Is the cleanup param optional?  It's not filled here.

File: ci/build.py [hunk 15]
```diff
@@ -37,17 +37,118 @@
...
+            stop_timeout = 3
+        for container in self.containers:
+            try:
+                container.stop(timeout=stop_timeout)
+                logging.info("☠: stopped container %s", trim_container_id(container.id))
+                container.remove()
+                logging.info("🚽: removed container %s", trim_container_id(container.id))
+            except Exception as e:
+                logging.exception(e)
+                #pass
```
Comment (KellenSunderland) [inline ci/build.py]: Please no commented code.

File: ci/build.py [hunk 16]
```diff
@@ -158,63 +246,142 @@ def default_ccache_dir() -> str:
...
     local_build_folder = buildir()
     # We need to create it first, otherwise it will be created by the docker daemon with root only permissions
     os.makedirs(local_build_folder, exist_ok=True)
     os.makedirs(local_ccache_dir, exist_ok=True)
     logging.info("Using ccache directory: %s", local_ccache_dir)
-    runlist = [docker_binary, 'run', '--rm', '-t',
+    docker_client = docker.from_env()
+    # Equivalent command
+    docker_cmd_list = [get_docker_binary(nvidia_runtime), 'run',
+               '--rm',
```
Comment (KellenSunderland) [inline ci/build.py]: Nit: indentation off.

File: ci/build.py [hunk 17]
```diff
@@ -37,17 +37,118 @@
...
+        self.containers.clear()
+        logging.info("Cleaning up containers finished.")
+
+    def __call__(self):
+        """Perform cleanup"""
+        self._cleanup_containers()
+
+
+
+def retry(ExceptionToCheck, tries=4, delay_s=1, backoff=2):
```
Comment (KellenSunderland) [inline ci/build.py]: arguments should be consistently lower case.

File: ci/build.py [hunk 18]
```diff
@@ -312,31 +521,37 @@ def use_cache():
...
-                          local_ccache_dir=args.ccache_dir)
+            command = ["/work/mxnet/ci/docker/runtime_functions.sh", "build_{}".format(platform)]
+            logging.info("No command specified, trying default build: %s", ' '.join(command))
+            ret = container_run(platform=platform, nvidia_runtime=args.nvidiadocker, shared_memory_size=shared_memory_size,
+                          command=command, docker_registry=args.docker_registry,
+                          local_ccache_dir=args.ccache_dir, cleanup=cleanup)
+
+        if ret != 0:
+            logging.critical("Execution of %s failed with status: %d", command, ret)
+            return(ret)
```
Comment (KellenSunderland) [inline ci/build.py]: redundant ()

File: ci/build.py [hunk 19]
```diff
@@ -37,17 +37,118 @@
...
+    def remove_container(self, container: docker.models.containers.Container):
+        assert isinstance(container, docker.models.containers.Container)
+        self.containers.remove(container)
+
+    def _cleanup_containers(self):
+        if self.containers:
+            logging.warning("Cleaning up containers")
+        else:
+            return
+        docker_client = docker.from_env()
```
Comment (KellenSunderland) [inline ci/build.py]: Where is docker_client used?

File: ci/build.py [hunk 20]
```diff
@@ -37,17 +37,118 @@
...
+        try:
+            stop_timeout = int(os.environ.get("DOCKER_STOP_TIMEOUT", self.docker_stop_timeout))
+        except Exception as e:
```
Comment (KellenSunderland) [inline ci/build.py]: e unused

File: ci/build.py [hunk 21]
```diff
@@ -158,63 +246,142 @@ def default_ccache_dir() -> str:
...
+                  nvidia_runtime: bool,
                   docker_registry: str,
                   shared_memory_size: str,
-                  local_ccache_dir: str,
                   command: List[str],
+                  local_ccache_dir: str,
+                  cleanup: Cleanup,
                   dry_run: bool = False,
-                  interactive: bool = False) -> str:
+                  interactive: bool = False) -> int:
```
Comment (KellenSunderland) [inline ci/build.py]: where is interactive used?

File: ci/build.py [hunk 22]
```diff
@@ -312,31 +521,37 @@ def use_cache():
...
-            container_run(platform=platform, docker_binary=docker_binary, shared_memory_size=shared_memory_size,
-                          command=cmd, docker_registry=args.docker_registry,
-                          local_ccache_dir=args.ccache_dir)
+            command = ["/work/mxnet/ci/docker/runtime_functions.sh", "build_{}".format(platform)]
+            logging.info("No command specified, trying default build: %s", ' '.join(command))
+            ret = container_run(platform=platform, nvidia_runtime=args.nvidiadocker, shared_memory_size=shared_memory_size,
+                          command=command, docker_registry=args.docker_registry,
+                          local_ccache_dir=args.ccache_dir, cleanup=cleanup)
+
+        if ret != 0:
```
Comment (KellenSunderland) [inline ci/build.py]: Where do we return 0?

File: ci/build.py (hunk 3 above)
Comment (larroy) [reply ci/build.py]: This works now as expected with the space, can be cut and pasted no problem.

File: ci/build.py (hunk 12 above)
Comment (larroy) [reply ci/build.py]: Why?

File: ci/build.py (hunk 3 above)
Comment (KellenSunderland) [reply ci/build.py]: Ok cool, I'll give it a shot again.

**Comment by aaronmarkham** [comment]
Can you provide any usage/testing info? 
What would be a good way to try this out locally: break/bite something and turn it into a zombie, and then kill it?

**Comment by larroy** [comment]
Depends on PR https://github.com/apache/incubator-mxnet/pull/12296

**Comment by larroy** [comment]
Made the change smaller with @KellenSunderland in a different PR: https://github.com/apache/incubator-mxnet/pull/12336
